aws_crt_common_ws/aws-crt-common/src/lib.rs

```diff
@@ -0,0 +1,596 @@
+    }
+}
+
+impl Drop for TracingAllocator {
+    fn drop(&mut self) {
+        // aws_mem_tracer_destroy() returns the wrapped allocator, but since in new()
+        // it's specified as the default_allocator, it doesn't need to be cleaned up.
+        unsafe {
+            let _ = aws_mem_tracer_destroy(self.c_allocator);
+        }
+    }
+}
+
+/// String type which wraps the internal aws_string structure. This type is immutable.
+#[non_exhaustive]
+pub struct AwsString {
+    c_str: *const CString,
+    owned: bool,
+}
+
+impl AwsString {
+    /// Creates a new String for a Rust std::string::String instance.
+    ///
+    /// Returns a result containing the new String or an int-based error code on failure.
+    pub fn new_from_std_string(
+        allocator: &dyn NativeResource<CAllocator>,
+        str: &String,
+    ) -> Result<AwsString, i32> {
+        return AwsString::new_from_array(allocator, str.as_bytes());
+    }
+
+    /// Creates a new String for a Rust str.
+    ///
+    /// Returns a result containing the new String or an int-based error code on failure.
+    pub fn new_from_str(
+        allocator: &dyn NativeResource<CAllocator>,
+        str: &str,
+    ) -> Result<AwsString, i32> {
+        let bytes = str.as_bytes();
+        return AwsString::new_from_array(allocator, bytes);
+    }
+
+    pub fn new_from_cstr(c_str: *const CString) -> AwsString {
+        AwsString {
+            c_str,
+            owned: false,
+        }
+    }
+
+    /// Creates a new string from an arbitrary array of utf-8 bytes.
+    ///
+    /// Returns a result containing the new String or an int-based error code on failure.
+    pub fn new_from_array(
+        allocator: &dyn NativeResource<CAllocator>,
+        str_data: &[u8],
+    ) -> Result<AwsString, i32> {
+        // assumes allocator is valid and str_data is not null.
+        // since the underlying allocation can fail, returns an error
+        // if aws_string_new_from_array() fails.
+        // The memory from this function is cleaned up in the destructor. If
+        // the destructor isn't run, the worse that will happen is a memory leak.
+        unsafe {
+            let c_str_ptr = aws_string_new_from_array(
+                allocator.get_native_resource(),
+                str_data.as_ptr(),
+                str_data.len(),
+            );
+            if c_str_ptr.is_null() {
+                return Err(last_error());
+            }
+
+            return Ok(AwsString {
+                c_str: c_str_ptr,
+                owned: true,
+            });
+        }
+    }
+
+    /// Returns the length of the string in bytes, not including any null terminators.
+    pub fn length(&self) -> usize {
+        // This is sound if you have an instance of this class.
+        unsafe {
+            return (*self.c_str).len;
+        }
+    }
+
+    /// Returns a pointer to an array of utf8 characters for the underlying String
+    pub fn bytes(&self) -> *const u8 {
```

Comment:
Idiomatically, this would be named `as_ptr()`

Reply:
e.g. https://doc.rust-lang.org/stable/std/string/struct.String.html#method.as_ptr

Reply:
You won't actually need this fn if you implement `Deref<Target = str>`

Reply:
Actually, this type does not contain `UTF-8` data, so you will not want to deref to `str`, instead you can deref to `[u8]`.